### PR TITLE
feat[integration-tests]: add verifier sync test

### PR DESCRIPTION
**Pull request by annieke**
**Description**
Still rough but marking this as ready for review since there is one passing test - would love feedback on the test structure and whether we should merge the dummy test while debugging the ERC20 test.

- Adds verifier to the docker-compose stack
- Adds integration test for syncing a verifier up to sequencer with dummy tx and ERC20 deployment + transfer

The structure of the new test:
1. uses existing stack (so the same sequencer every test, which shouldn't be a problem if sync tests pass)
2. spins up a new verifier per test case

This means each test takes quite long. The dummy tx test took 14.5s and the ERC20 one took 24.5s - feedback on how to speed this up welcome!

**Additional context**

For the failing ERC20 test, the test still fails with only the contract deployment and not the transfer.

**Metadata**
- OP-533
- https://github.com/ethereum-optimism/optimism/issues/604

### CODE REVIEW

File: ops/verifier.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    env_file:
+      - ./envs/geth.env
+    environment:
+        ETH1_HTTP: http://l1_chain:8545
+        ROLLUP_STATE_DUMP_PATH: http://deployer:8081/state-dump.latest.json
+        # used for getting the addresses
+        URL: http://deployer:8081/addresses.json
+        # connecting to the DTL
+        # TODO(annieke): should this be different for verifier?
+        ROLLUP_CLIENT_HTTP: http://dtl:7878
```
Comment (annieke) [inline ops/verifier.yml:23]: @tynes i haven't gotten this running yet, would connecting the verifier to the same dtl be a problem?

File: ops/README.md [hunk 2]
```diff
@@ -28,6 +28,14 @@ docker-compose \
     up --build --detach
 ```
 
+Do the same to run a verifier along side the regular stack.
+```
+docker-compose \
+    -f docker-compose.yml \
+    -f verifier.yml \
```
Comment (tynes) [inline ops/README.md:35]: Nit: follow the style of the `docker-compose-` prefix to keep things a more unified style, see the metrics example above

File: ops/verifier.yml (hunk 1 above)
Comment (tynes) [reply ops/verifier.yml:23]: I think it should work

Comment (tynes) [reply ops/verifier.yml:23]: When https://github.com/ethereum-optimism/optimism/pull/552 lands then I know it will definitely work. Pre #552 then it should work as long as the DTL default backend is L1

**Review by gakonst — CHANGES_REQUESTED** [review]
@annieke instead of creating a separate file, I recommend putting it in the same file and setting the `replicas: 0` setting, similar to how we do with the `integration_tests` image: https://github.com/ethereum-optimism/optimism/blob/develop/ops/docker-compose.yml#L134-L135.

Then, you could bring it up with `docker-compose up --scale verifier=1`.

**Comment by annieke** [comment]
more context for the mismatched stateroot: i'm manually comparing the two blocks and they contain the same transaction, which makes me think this is a legitimate mismatch. thoughts @karlfloersch ?

### sequencer block with transaction
```json
{
   "jsonrpc":"2.0",
   "id":1,
   "result":{
      "difficulty":"0x2",
      "extraData":"0xd98301090a846765746889676f312e31342e3135856c696e757800000000000057bb4fd04e3304ec3d3e3001d25a1cef856d3077e27f62685a29a1f8729033b42220f9ef93a16984c1d50b6f0fff37d1db0205afe687085254730142239617a800",
      "gasLimit":"0x895440",
      "gasUsed":"0x8e1e1",
      "hash":"0xdea8a1bb9f6d3798b629981e07856e0b25ce7fb5e20b7e9a75f3361e21deac43",
      "logsBloom":"0x00000000000000000000000000000000000800000000000000040000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000008000004000000000010000000000000000000000400000040000000000000000100008000000000800000000000000010000000000000000000000000000000000000000000000000002000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000042000000200000000000000000000000002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "miner":"0x0000000000000000000000000000000000000000",
      "mixHash":"0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce":"0x0000000000000000",
      "number":"0x3",
      "parentHash":"0xd208c21184a9f7c1d38f76bf375831f0b2433d85c6bc451d70fe5f567cd86204",
      "receiptsRoot":"0x6ac57aa4c7e87e5761883bc21497d77e21a838ab12760c7b5257b6b171eb9f63",
      "sha3Uncles":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
      "size":"0x30e",
      "stateRoot":"0xb27fa216bda5d3a04fc0e0297409aeef40b52811523ccd78261a1fed740bb08d",
      "timestamp":"0x60ac4d24",
      "totalDifficulty":"0x7",
      "transactions":[
         {
            "blockHash":"0xdea8a1bb9f6d3798b629981e07856e0b25ce7fb5e20b7e9a75f3361e21deac43",
            "blockNumber":"0x3",
            "from":"0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
            "gas":"0x5208",
            "gasPrice":"0x3b9aca00",
            "hash":"0x156dee9dc95d323d0b22c0f175fdc9b4a059a371c058a4ffcdbe2d359501a2b4",
            "input":"0xa9059cbb000000000000000000000000050acfa1e62f6fb0e6de77246a5b41aa41d4a1e70000000000000000000000000000000000000000000000000000000000000064",
            "nonce":"0x1",
            "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3",
            "transactionIndex":"0x0",
            "value":"0x0",
            "v":"0x36c",
            "r":"0x871a64832d904e716b84c9a530651a908ad69d0534ed51bc3a882b95fb91c223",
            "s":"0x3923ddee29d88351394951d2a342ff50de48b4dd9e67ce79689c20a26aff8478",
            "queueOrigin":"sequencer",
            "txType":"EIP155",
            "l1TxOrigin":null,
            "l1BlockNumber":"0x2b",
            "l1Timestamp":"0x60ac4d24",
            "index":"0x2",
            "queueIndex":null,
            "rawTransaction":"0xf8aa01843b9aca00825208945fbdb2315678afecb367f032d93f642f64180aa380b844a9059cbb000000000000000000000000050acfa1e62f6fb0e6de77246a5b41aa41d4a1e7000000000000000000000000000000000000000000000000000000000000006482036ca0871a64832d904e716b84c9a530651a908ad69d0534ed51bc3a882b95fb91c223a03923ddee29d88351394951d2a342ff50de48b4dd9e67ce79689c20a26aff8478"
         }
      ],
      "transactionsRoot":"0xb466f6e0d651f976c7e91a0cd71812f81e1445755237002c3b1d3f0f0326505e",
      "uncles":[
         
      ]
   }
}
```

### verifier block with transaction

```json
{
   "jsonrpc":"2.0",
   "id":1,
   "result":{
      "difficulty":"0x2",
      "extraData":"0xd98301090a846765746889676f312e31342e3135856c696e7578000000000000d48b533732a4b3a3e4f1d01e6fd3f9bffd2fe26a1e2fa6222faa5abf7db5460f400cebb359c6137142191ad893ca85d810d7659997dcc728eb19ad62fa8099f401",
      "gasLimit":"0x895440",
      "gasUsed":"0x8e1e1",
      "hash":"0xbbc30e5a401df7dd8a19a4e86d262ea8714bffebd255b83bd4a3386d99a1c658",
      "logsBloom":"0x00000000000000000000000000000000000800000000000000040000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000008000004000000000010000000000000000000000400000040000000000000000100008000000000800000000000000010000000000000000000000000000000000000000000000000002000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000042000000200000000000000000000000002000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "miner":"0x0000000000000000000000000000000000000000",
      "mixHash":"0x0000000000000000000000000000000000000000000000000000000000000000",
      "nonce":"0x0000000000000000",
      "number":"0x3",
      "parentHash":"0x65b172d3cde34eb0abb1462d4ef35b7db2771cc76c10c97d3c84fc593fd12b32",
      "receiptsRoot":"0x6ac57aa4c7e87e5761883bc21497d77e21a838ab12760c7b5257b6b171eb9f63",
      "sha3Uncles":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
      "size":"0x30e",
      "stateRoot":"0x707d07837867afb72216bbc9cb4da9c34cf00f7d6be3443c456253b3304fe8d1",
      "timestamp":"0x60ac4d24",
      "totalDifficulty":"0x7",
      "transactions":[
         {
            "blockHash":"0xbbc30e5a401df7dd8a19a4e86d262ea8714bffebd255b83bd4a3386d99a1c658",
            "blockNumber":"0x3",
            "from":"0x517ab03785a2cc5379429d537c203f69b97b7630",
            "gas":"0x5208",
            "gasPrice":"0x3b9aca00",
            "hash":"0x6098ac748b225c61bc98e85cbe75429c7d73b81dd656e5831eb19958bea78481",
            "input":"0xa9059cbb000000000000000000000000050acfa1e62f6fb0e6de77246a5b41aa41d4a1e70000000000000000000000000000000000000000000000000000000000000064",
            "nonce":"0x1",
            "to":"0x5fbdb2315678afecb367f032d93f642f64180aa3",
            "transactionIndex":"0x0",
            "value":"0x0",
            "v":"0x3d7",
            "r":"0x871a64832d904e716b84c9a530651a908ad69d0534ed51bc3a882b95fb91c223",
            "s":"0x3923ddee29d88351394951d2a342ff50de48b4dd9e67ce79689c20a26aff8478",
            "queueOrigin":"sequencer",
            "txType":"EIP155",
            "l1TxOrigin":null,
            "l1BlockNumber":"0x2b",
            "l1Timestamp":"0x60ac4d24",
            "index":"0x2",
            "queueIndex":null,
            "rawTransaction":"0xf8aa01843b9aca00825208945fbdb2315678afecb367f032d93f642f64180aa380b844a9059cbb000000000000000000000000050acfa1e62f6fb0e6de77246a5b41aa41d4a1e7000000000000000000000000000000000000000000000000000000000000006482036ca0871a64832d904e716b84c9a530651a908ad69d0534ed51bc3a882b95fb91c223a03923ddee29d88351394951d2a342ff50de48b4dd9e67ce79689c20a26aff8478"
         }
      ],
      "transactionsRoot":"0x48e9415c1e084c98f091a9d37e79a83c41206b5d734f02545ae9a88887933807",
      "uncles":[
         
      ]
   }
}
```

File: integration-tests/test/sync-verifier.spec.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  const provider = injectL2Context(l2Provider)
+
+  /** Helper functions **/
+
+  const waitForBatchSubmission = async (totalElementsBefore: BigNumber) => {
+    // Wait for batch submission to happen by watching the CTC
+    let totalElementsAfter = (await env.ctc.getTotalElements()) as BigNumber
+    while (totalElementsBefore.eq(totalElementsAfter)) {
+      await sleep(500)
+      console.log(
```
Comment (annieke) [inline integration-tests/test/sync-verifier.spec.ts:24]: note: i'll remove all these `console.log`s if we decide to merge the dummy tx test first; leaving these here in case someone else wants to pull and debug

File: integration-tests/test/sync-verifier.spec.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      other = Wallet.createRandom().connect(ethers.provider)
+      Factory__ERC20 = await ethers.getContractFactory('ERC20', wallet)
+    })
+
+    afterEach(async () => {
+      await verifier.stop('verifier')
+      await verifier.rm()
+    })
+
+    it('should sync dummy transaction', async () => {
```
Comment (annieke) [inline integration-tests/test/sync-verifier.spec.ts:90]: this test currently passes!

File: integration-tests/test/sync-verifier.spec.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      const latestVerifierBlock = await startAndSyncVerifier(
+        latestSequencerBlock.number
+      )
+
+      expect(latestVerifierBlock.stateRoot).to.eq(
+        latestSequencerBlock.stateRoot
+      )
+    })
+
+    it('should sync ERC20 deployment and transfer', async () => {
```
Comment (annieke) [inline integration-tests/test/sync-verifier.spec.ts:120]: this test fails when it's deploy and transfer, also when it's only deploy without the transfer. i posted a comment with the failing block and verified it should be the same block therefore a real bug

File: integration-tests/test/sync-verifier.spec.ts (hunk 4 above)
Comment (gakonst) [reply integration-tests/test/sync-verifier.spec.ts:88]: This test halts for me at `waiting for new verifier blocks, current:  1 latest 2` when run with `it.only`

Comment (annieke) [reply integration-tests/test/sync-verifier.spec.ts:88]: @gakonst did you quit a previous test or had a verifier running locally before you started? i just remembered i ran into that before i removed the verifier image after every test; instead of restarting to sync the verifier just sits at the previous block

**Comment by tynes** [comment]
A good way to debug is to make sure that the genesis state roots match like so:

```
ETH_RPC_URL=... seth block 0 stateRoot
```

Where `ETH_RPC_URL` is set to the HTTP endpoint of the verifier/sequencer

Another way to debug is look at the gasUsed - sometimes the gas used will greatly differ and this can give a hint as to what went wrong, ie how far the execution got relative to each other. Make sure that the `rawTransaction` hex matches exactly because this was the cause of the last bug. Use the scripts here to compare verifiers and sequencers:
- https://github.com/ethereum-optimism/scripts/blob/main/scripts/verifier/compare-verifier-sequencer.js
- https://github.com/ethereum-optimism/scripts/blob/main/scripts/verifier/diff-verifier-sequencer-state-dump.js

File: ops/docker/Dockerfile.geth [hunk 6]
```diff
@@ -12,6 +12,6 @@ FROM alpine:latest
 RUN apk add --no-cache ca-certificates jq curl
 COPY --from=builder /go-ethereum/build/bin/geth /usr/local/bin/
 
-EXPOSE 8545 8546 8547
+EXPOSE 8545 8546 8547 8548
```
Comment (tynes) [inline ops/docker/Dockerfile.geth:15]: We shouldn't add another port here - adding the port to the docker compose file should be fine

File: ops/scripts/check-sequencer.sh [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+set -e
+
```
Comment (tynes) [inline ops/scripts/check-sequencer.sh:3]: Where is this script used?

File: integration-tests/tsconfig.json [hunk 8]
```diff
@@ -3,6 +3,6 @@
   "compilerOptions": {
       "resolveJsonModule": true
   },
-  "include": ["./test"],
```
Comment (tynes) [inline integration-tests/tsconfig.json:6]: This isn't scalable, should use a glob match like `sync-tests/*`

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      await sleep(500)
+      logs = await verifier.logs()
+    }
+
+    const verifierProvider = injectL2Context(
+      new providers.JsonRpcProvider('http://localhost:8547')
+    )
+
+    // Wait until verifier has caught up to the sequencer
+    let latestVerifierBlock = (await verifierProvider.getBlock('latest')) as any
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:45]: Is there a type in `core-utils` that represents the L2 block? At one point I think there was one in the batch submitter

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+    // Wait for verifier to be looping
+    let logs = await verifier.logs()
+    while (!logs.out.includes('Starting Sequencer Loop')) {
+      await sleep(500)
+      logs = await verifier.logs()
+    }
+
+    const verifierProvider = injectL2Context(
+      new providers.JsonRpcProvider('http://localhost:8547')
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:41]: This http endpoint needs to come from `OptimismEnv`

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+describe('Syncing a verifier', () => {
+  let env: OptimismEnv
+  let wallet: Wallet
+  let verifier: DockerComposeNetwork
+
+  const provider = injectL2Context(l2Provider)
+
+  /* Helper functions */
+
+  const waitForBatchSubmission = async (totalElementsBefore: BigNumber) => {
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:18]: In an ideal world, we do not spin up the batch submitter and submit the batches manually using JS. This would cut out the extra dependency of needing to have a whole other codepath and the complexity that comes with it, like this function

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      }
+      const result = await wallet.sendTransaction(tx)
+      await result.wait()
+
+      const totalElementsAfter = await waitForBatchSubmission(
+        totalElementsBefore
+      )
+      expect(totalElementsAfter.gt(totalElementsAfter))
+
+      const latestSequencerBlock = (await provider.getBlock('latest')) as any
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:83]: You should be explicit around getting blocks at the same height instead of latest to prevent race conditions

File: .github/workflows/sync-tests.yml [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+      - name: Bring the stack up
+        working-directory: ./ops
+        run: docker-compose up -d && ./scripts/wait-for-sequencer.sh
+      
+      - name: Run the sync tests
+        working-directory: ./integration-tests
+        run: |
+          yarn
+          yarn test:sync
```
Comment (tynes) [inline .github/workflows/sync-tests.yml:47]: Is there any reason that this is a different workflow than the other integration tests? Ideally the images are built once and then used twice in parallel

File: .github/workflows/sync-tests.yml [hunk 14]
```diff
@@ -0,0 +1,47 @@
+name: sync-tests
+
+on:
+  push:
+    branches:
+      - 'master'
+      - 'develop'
+      - '*rc'
+      - 'regenesis/*'
```
Comment (tynes) [inline .github/workflows/sync-tests.yml:9]: It looks like we have branch creep

File: integration-tests/package.json [hunk 15]
```diff
@@ -13,6 +13,7 @@
     "build:contracts": "hardhat compile",
     "build:contracts:ovm": "hardhat compile --network optimism",
     "test:integration": "hardhat --network optimism test",
+    "test:sync": "hardhat --network optimism test sync-tests/*.spec.ts --no-compile",
```
Comment (tynes) [inline integration-tests/package.json:16]: I feel like we need sync tests that have a dummy db that is passed to geth (upstream geth in L1 mode because hardhat doesn't allow for this) and then we sync from that deterministically. Perhaps we can call these verifier tests? We need to do the exact same logic for replicas so maybe network tests? Since its a network of nodes?

File: ops/scripts/check-sequencer.sh (hunk 7 above)
Comment (annieke) [reply ops/scripts/check-sequencer.sh:3]: oo this needs to be deleted, will do that

File: integration-tests/package.json (hunk 15 above)
Comment (annieke) [reply integration-tests/package.json:16]: ^can rename to network tests!

File: .github/workflows/sync-tests.yml (hunk 13 above)
Comment (annieke) [reply .github/workflows/sync-tests.yml:47]: ^hm i can look into this! just know we need a fresh sequencer for this set of tests

Comment (gakonst) [reply .github/workflows/sync-tests.yml:47]: I had started working on that [here](https://github.com/ethereum-optimism/optimism/pull/891/files#diff-82454b2fc887e9985b9348b8f3bca03d2f839645a4c3ba1f3f850014d7da5c8bR41), agreed with Mark but I think it's better to do this in another PR, in the interest of saving Annie's time

Tl;dr: You'd have 1 job which builds the services, saves the docker images to a tar and then uploads the artifacts. Then in the tests job you'd download the artifacts, load the docker images and run the tests.

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 9 above)
Comment (gakonst) [reply integration-tests/sync-tests/sync-verifier.spec.ts:45]: There is not, we could add it in a separate PR as it warrants a bigger refactor probably.

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 11 above)
Comment (gakonst) [reply integration-tests/sync-tests/sync-verifier.spec.ts:18]: Given that these are integration tests checking the relation between the sequencer's sync service, the dtl, the verifier and the batch submitter, I think it's better as is actually

File: integration-tests/tsconfig.json [hunk 16]
```diff
@@ -3,6 +3,6 @@
...
+  "include": ["./test", "sync-tests/sync-verifier.spec.ts"],
```
Comment (gakonst) [inline integration-tests/tsconfig.json:6]:
```suggestion
  "include": ["./test", "sync-tests/*.ts"],
```

**Review by gakonst — APPROVED** [review]
LGTM in terms of content. I'm fine with merging the PR as-is and merging the workflows in a separate PR, since it's docker/CI heavy.

File: .github/workflows/sync-tests.yml (hunk 14 above)
Comment (gakonst) [reply .github/workflows/sync-tests.yml:9]: @snario we've agreed to move away from `*rc` right? The other 3 are correct otherwise @tynes

File: integration-tests/package.json (hunk 15 above)
Comment (gakonst) [reply integration-tests/package.json:16]: I don't think we necessarily need a pre-instantiated db as things are reasonably deterministic in our current envs? Think PR is fine rn otherwise

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 11 above)
Comment (tynes) [reply integration-tests/sync-tests/sync-verifier.spec.ts:18]: I'm fine with leaving as is, but I think the more controlled the test, the better. This isn't a batch-submitter test, this is a verifier test. It should test the bare minimal stuff to make sure the verifier works.

**Review by tynes — CHANGES_REQUESTED** [review]
Do not merge with with the changed dockerfile

File: integration-tests/package.json (hunk 15 above)
Comment (tynes) [reply integration-tests/package.json:16]: That was more of a longer term comment, sorry for the confusion. In the long term we need sync tests that basically sync thousands of blocks where we can observe things like performance and state

Comment (annieke) [reply integration-tests/package.json:16]: ^opted to rename those later since sync is clearer for this context :)

File: ops/docker/Dockerfile.geth [hunk 17]
```diff
@@ -12,6 +12,6 @@ FROM alpine:latest
 RUN apk add --no-cache ca-certificates jq curl
 COPY --from=builder /go-ethereum/build/bin/geth /usr/local/bin/
 
-EXPOSE 8545 8546 8547
+EXPOSE 8545 8546 8547
```
Comment (tynes) [inline ops/docker/Dockerfile.geth:15]: There is an extra whitespace here

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    env = await OptimismEnv.new()
+    wallet = env.l2Wallet
+  })
+
+  describe('Basic transactions', () => {
+    afterEach(async () => {
+      await verifier.stop('verifier')
+      await verifier.rm()
+    })
+
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:64]: Why does `afterEach` remove the verifier but there is no `beforeEach` to bring it up? Shouldn't this be a single `before` to bring it up (decouple starting and syncing) and just start and then a single before to clean up?

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      const totalElementsAfter = await waitForBatchSubmission(
+        totalElementsBefore
+      )
+      expect(totalElementsAfter.gt(totalElementsAfter))
+
+      const latestSequencerBlock = (await sequencerProvider.getBlock(
+        'latest'
+      )) as any
+
+      const matchingVerifierBlock = (await startAndSyncVerifier(
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:87]: I think its best to split up starting the verifier and syncing the verifier - that will make this code more flexible longer term

File: integration-tests/sync-tests/sync-verifier.spec.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+    await verifier.up({ commandOptions: ['--scale', 'verifier=1'] })
+
+    // Wait for verifier to be looping
+    let logs = await verifier.logs()
+    while (!logs.out.includes('Starting Sequencer Loop')) {
+      await sleep(500)
+      logs = await verifier.logs()
+    }
+
+    const provider = injectL2Context(verifierProvider)
```
Comment (tynes) [inline integration-tests/sync-tests/sync-verifier.spec.ts:42]: The verifier provider can be decoupled and instantiated at the very top of the file just like the `sequencerProvider`

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 18 above)
Comment (annieke) [reply integration-tests/sync-tests/sync-verifier.spec.ts:64]: i followed the structure described in https://github.com/ethereum-optimism/optimism/issues/604 which specified spinning up the verifier after making transactions. i haven't looked into distinctions between the two approaches yet, just imagined this test as "starting with a fresh copy every time"

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 19 above)
Comment (annieke) [reply integration-tests/sync-tests/sync-verifier.spec.ts:87]: same reasoning as above^

File: integration-tests/sync-tests/sync-verifier.spec.ts (hunk 20 above)
Comment (annieke) [reply integration-tests/sync-tests/sync-verifier.spec.ts:42]: same rationale behind the "new synced verifier" per test as described in the issue! i realize this does slow tests down, but before we have a bunch of those and if they can be parallelized, i think integration tests will still be the bottleneck here. would rather have this running and gradually iterate instead of get this final in the first round

Comment (tynes) [reply integration-tests/sync-tests/sync-verifier.spec.ts:42]: What I am suggesting is along the lines of:

```js
function startVerifier() {
    verifier = new DockerComposeNetwork(['verifier'])
    await verifier.up({ commandOptions: ['--scale', 'verifier=1'] })

    // Wait for verifier to be looping
    let logs = await verifier.logs()
    while (!logs.out.includes('Starting Sequencer Loop')) {
      await sleep(500)
      logs = await verifier.logs()
    }
}

function syncVerifier(provider) {
    // Wait until verifier has caught up to the sequencer
    let latestVerifierBlock = (await provider.getBlock('latest')) as any
    while (latestVerifierBlock.number < sequencerBlockNumber) {
      await sleep(500)
      latestVerifierBlock = (await provider.getBlock('latest')) as any
    }

    return provider.getBlock(sequencerBlockNumber)
}
```

You can still spin up the verifier after making transactions with this abstraction. This will make the code easier to maintain longer term because the abstractions are cleaner. You no longer need to sync the verifier when bringing it up. This abstraction split will eventually need to happen as more tests are added to this test suite

**Review by tynes — APPROVED** [review]
Looks good to me - I cannot seem to find evidence of the tests running in CI, just making sure that the tests are running as expected

**Comment by annieke** [comment]
@tynes i'm planning to add this test to CI in a follow up PR per this suggestion https://github.com/ethereum-optimism/optimism/pull/913#pullrequestreview-668742610; will continue the work here https://github.com/ethereum-optimism/optimism/pull/891 if that sounds good!
